pkg/virt-handler/dmetrics-manager/dmetrics-manager.go

```diff
@@ -0,0 +1,120 @@
+	virtioserial "kubevirt.io/kubevirt/pkg/downwardmetrics/virtio-serial"
+	cmdclient "kubevirt.io/kubevirt/pkg/virt-handler/cmd-client"
+	"kubevirt.io/kubevirt/pkg/virt-handler/isolation"
+)
+
+func NewDownwardMetricsManager(nodeName string, virtSharedDir string) *DownwardMetricsManager {
+	return &DownwardMetricsManager{
+		done:          false,
+		nodeName:      nodeName,
+		virtSharedDir: virtSharedDir,
+		stopServer:    make(map[string]context.CancelFunc),
+	}
+}
+
+// DownwardMetricsManager controls the lifetime of the DownwardMetrics servers.
+// Each server is tied to the lifetime of the VMI and DownwardMetricsManager itself.
+type DownwardMetricsManager struct {
+	lock          sync.Mutex
+	done          bool
+	nodeName      string
+	virtSharedDir string
+	stopServer    map[string]context.CancelFunc
+}
+
+// Run blocks until stopCh is closed. When done, it stops all remaining
+// running DownwardMetrics servers.
+func (m *DownwardMetricsManager) Run(stopCh chan struct{}) {
+	defer m.stop()
+	<-stopCh
+}
+
+func (m *DownwardMetricsManager) stop() {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	m.done = true
+
+	// Stop all DownwardMetrics servers
+	for domainName, stopServerFn := range m.stopServer {
+		stopServerFn()
+		delete(m.stopServer, domainName)
+	}
+}
+
+// StopServer removes the VMI name from the list of served VMs a
+// nd stop the DownwardMetrics server if necessary
+func (m *DownwardMetricsManager) StopServer(vmi *v1.VirtualMachineInstance) {
+	if !downwardmetrics.HasDevice(&vmi.Spec) {
+		return
+	}
+
+	domainName := fmt.Sprintf("%s/%s", vmi.GetNamespace(), vmi.GetName())
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if m.done {
+		return
+	}
+
+	// Even it's not strictly required for stopping the server,
+	// since the server will stop when the VM closes the unix socket,
+	// we cancel the context to avoid leaking it
+	m.stopServer[domainName]()
+	delete(m.stopServer, domainName)
+}
+
+// StartServer start a new DownwardMetrics server if the VM request it and is not already started
+func (m *DownwardMetricsManager) StartServer(vmi *v1.VirtualMachineInstance) error {
+	if !downwardmetrics.HasDevice(&vmi.Spec) {
+		return nil
+	}
+
+	if !vmi.IsRunning() {
+		return nil
+	}
+
+	domainName := fmt.Sprintf("%s/%s", vmi.GetNamespace(), vmi.GetName())
+
+	m.lock.Lock()
+	defer m.lock.Unlock()
+	if m.done {
+		return nil
+	}
+
+	if _, alreadyStarted := m.stopServer[domainName]; alreadyStarted {
+		return nil
+	}
+
+	launcherSocketPath, err := cmdclient.FindSocketOnHost(vmi)
+	if err != nil {
+		return fmt.Errorf("failed to get the launcher socket for VMI [%s], error: %v", vmi.GetName(), err)
+	}
+
+	socketDetector := isolation.NewSocketBasedIsolationDetector(m.virtSharedDir)
+	res, err := socketDetector.Detect(vmi)
+	if err != nil {
+		return fmt.Errorf("failed to detect root directory of the vmi pod for VMI [%s], error: %v", vmi.GetName(), err)
+	}
+
+	channelPath := downwardmetrics.ChannelSocketPathOnHost(res.Pid())
```

Comment:
I think requiring the `IsolationResult` as a function parameter would be much simpler for you. It would also allow you to hold the lock less which is always good.

Reply:
done, thanks